GVFS/GVFS.Common/Git/GitAuthentication.cs

```diff
@@ -45,6 +46,16 @@ public void ConfirmCredentialsWorked(string usedCredential)
                 {

```

Comment:
I realize this is existing but saw it was curious.  Why do we action only if (usedCredential == this.cachedAuthString).  Since this function is saying we've confirmed this credential works could just always save usedCredential?  Is there a case I'm missing here?

Reply:
The line I'm referencing got cut off, but I was wondering about 

`                if (usedCredential == this.cachedAuthString)
`

in ConfirmCredentialsWorked.  Why can't we just always use the confirmed string?

Reply:
I'm also not entirely sure why this check is done (this is place when code comments would have been useful!)
I didn't want to change things however because it seems to be some defence against cases when there are two different cached auth strings, and avoiding a reset of the backoff mechanism erroneously.

Still not entirely sure how or if such a situation could occur however. Perhaps it was a check that wasn't really needed. Thoughts?

Reply:
@mjcheetham I totally agree, it's hard to know the thought process without comments.  

To me the correct flow would be take the key that worked, break it apart and store.  I would think if GCM failed and got called again we may get different sets, but even still why not just take the working one?  If you need to get this PR in now maybe keep as is.  If it's going out in the next release I think it's worth fixing up to what we think the right behavior should be.

Reply:
@jeschu1 so I think I know why we're checking first that the credentials we used are the same as the ones we have cached.

If the actual stored credentials in the credstore/keychain have been replaced with new ones (outside of our knowledge), but the ones we last returned in a call to `TryGetCredentials` (and we have cached) are different, we don't want to tell Git to erase 'the credentials for the repo' as this might erase the now good ones. Note that we don't actually tell Git "hey, credential X is bad" we can only say "the credentials you know about are bad".

Reply:
Of course this is all a problem of potentially having a stale cached value, and knowing when the cache is stale or if the stored credentials are bad. Git was never designed to be used this way.. all calls to `git credential fill` should be immediately followed by either `approve` or `reject`, and then `git-credential` should only be called by one process at a time per remote URL.

Interleaving calls to `fill` and `reject` for the same remote URL on a single system will lead to problems. There's an assumed static state after calling `fill` that no one has touched the credentials we just got back.

If `reject` also took the actual credential we want to say didn't work, then the helpers would be able to check if that's the actual credential value they are being asked to erase.

Reply:
@jeschu1 I've added a small comment here about not resetting the backoff for a non-cached credential string. I don't want to change this too much at this point.

However you got my thinking if new GCM could help prevent this sort of problem in the future. I think we can.
Git itself will call `reject` passing the username and password it got from the `fill` call, meaning GCM can be smarter about how it handles requests to delete.

_**If**_ the username and password were specified on a call to `reject/erase`, then we can double check that the stored credential we have matches the user/pass we were given, and not erase it if it's different from the one we were told about. If we're not given a user/pass then just delete the credential without checking if the user/pass match the stored values.

Since Git is always consistent in what it sends to `reject` (what it got back from `fill`) the normal path here for Git deleting an invalid cred is still OK. Older VFS for Git's don't send any username or password so GCM will continue to delete any old credential values for the given key.

One day we might be able to remove the check in 'Revoke()' and just always pass through the credential strings, ignoring checking the cached version, but we'll always need to check the cached value to ensure we're not resetting the backoff settings. :(

Reply:
@mjcheetham I think I see the point here.  So if another process replaces the credentials we wouldn't necessarily want to store/replace ours.  I guess that's fair enough.  The flow is a bit confusing, but it predates your work here.  Thanks for reasoning that out.

Reply:
@mjcheetham  @jeschu1, I did some digging and this question was raised in the PR that initially added this check ("PR 223211: Improve GitAuthentication retry logic in multithreaded scenarios" in the old Azure Repos GVFS repo):

>Saeed Noursalehi
6/5/2017
In what scenario would this condition not be true?
>
>Forest Willard
6/5/2017
If a thread starts a long request as a PAT expires, then it will succeed with that PAT, but another thread may have already experienced a failure and changed the PAT.
>
>Saeed Noursalehi
6/5/2017
Removing this if doesn't cause any test failures. 

I have not tried removing the check but based on Saeed's comment I am assuming there is a test that fails without this check?

There was a lengthier discussion where the check was added in the Revoke/Reject method:

>Saeed Noursalehi
6/5/2017
Do we actually need this check? As I step through the race scenarios, I think the check below for "cachedAuthString != null" would also handle the race.
>
>The actual bug was that in the old code, the else path in this method would set authRetryBackoff to DateTime.MaxValue. Now that that line is gone, I think all of the rest of the logic works OK without having to pass around this extra usedCredential parameter.
>
>Thoughts?
>
>Forest Willard
6/5/2017
This handles a different race than the bug. If I remove this line, I'm pretty darn sure "TwoThreadsInterleavingFailuresShouldntStompASuccess" will start failing. The effects of it failing aren't so bad, though. It just means that we would create one extra PAT when we don't need to, given everything returns success. If two threads grab a failing PAT, then one retries before the second can fully fail, the second will revoke the good PAT the first retried with. Then we will renew it later.
>
>Actually, I think it'd be possible to get failed requests out of this race, but it's highly unlikely. When the second thread revokes the good PAT, there could be a brief period where the first thread hasn't reset the backoff counter, and anything else would fail without trying.
>
>Saeed Noursalehi
6/5/2017
I don't really follow what the race is, or how this check fixes it. The effect of this check here is that this method returns true and the caller will retry. If you remove this check, cachedAuthString will be null, and DateTime.Now will be less than the backoff time.
>
>Unless... if the first thread actually succeeds in recreating the PAT before the second thread even got to rejecting the old one, then cachedAuthString won't be null, and we revoke again. Is that the race?
>
>Forest Willard
6/5/2017
Yup, that's the race.
>
>Saeed Noursalehi
6/5/2017
OK makes sense, and I confirmed that removing this code causes a unit test failure, so that's good.

Reply:
I've added a new [issue](https://github.com/Microsoft/Git-Credential-Manager-Core/issues/24) (and PR) on GCM Core to hopefully avoid this sort of problem (or at least push co-ordination of deleting credentials back to GCM). VFS will still need to do this sort of check however to update/reset its backoff.